base/operators.jl

```diff
@@ -111,9 +111,10 @@ Use [`isequal`](@ref) or [`===`](@ref) to always get a `Bool` result.
 New numeric types should implement this function for two arguments of the new type, and
 handle comparison to other types via promotion rules where possible.
 
-[`isequal`](@ref) falls back to `==`, so new methods of `==` will be used by the
-[`Dict`](@ref) type to compare keys. If your type will be used as a dictionary key, it
-should therefore also implement [`hash`](@ref).
+Equality and hashing are intimately related; two values that are considered [`isequal`](@ref) **must**
+have the same [`hash`](@ref). If a type customizes the behavior of `==` and/or [`isequal`](@ref),
```

Comment:
As I understand it, the defining difference between `isequal`  and `==` is that implementing the latter doesn't require `hash`. The invariant is that `isequal(x, y)` implies `hash(x) == hash(y)`. It's unfortunate that `isequal` falls back to `==` by default, but that shouldn't mean that the contract of `==` is strengthened.

Reply:
The most important point here is that it's _not just dictionary lookups **you** might perform_, because Julia itself rightfully assumes this invariant holds in many places.

I indicate that `isequal` is the important part of the invariant in the first sentence.  My second sentence is also correct: implementing _any_ combination of `==`, `isequal`, or both requires implementing `hash`.

I'm not sure what you want me to change here.

Reply:
> The most important point here is that it's not just dictionary lookups you might perform, because Julia itself rightfully assumes this invariant holds in many places.

Yes, I understand that. My example with `dictionary key` was just to indicate that I don't think saying that if you customize `==`, you HAVE to implement `hash` too to fulfill the contract of `==`.

> I indicate that isequal is the important part of the invariant in the first sentence. My second sentence is also correct: implementing any combination of ==, isequal, or both requires implementing hash.

I don't think that's correct. `isequal` & `hash` will silently give wrong results if you customize `==`, but that doesn't mean that the implementation of `==` is incorrect. As is, the text in your PR implies that this is the case, which would strengthen the contract of `==`  and cause existing implementations (that don't require hash-equality for their usecase) to be incorrect.

Specifically, it's this reading:

```
If a type customizes the behavior of `==` [...], then [`hash`](@ref) must be similarly implemented. 
```

that I think is not good.

Reply:
You keep eliding the important parts with `[...]` to make this say things I didn't intend.  It's just three sentences, and I carefully wrote all three.  But I do agree it's helpful to maintain the information that `isequal` falls back to `==` and, sure, that the "similarly implemented" is _just_ about whatever behavior `isequal` ended up having.

Reply:
> You keep eliding the important parts with [...] to make this say things I didn't intend

Well, yeah, that's the point - the intention can't be guessed, and either reading is valid :) As a docstring though, I'd much rather assume an unintentionally "bad" reading and seeing if that would still convey the same meaning. If it doesn't, the docstring is ambiguous.